### DISCUSSION TITLE
Feature request for CLI: Seed storage files

### DISCUSSION THREAD

**Discussion by joeally** [supabase/supabase#12390] [Feature Requests] [answered]
Every time I run `db reset` all data is lost. On the database side this can be remedied with `seed.sql` however there doesn't appear to be anyway to do this for storage objects. This makes for a fairly frustrating dev experience where I have to recreate the buckets and reupload the same files every time I want to test a migration locally.

It would be amazing if you could define some seed files which would be uploaded every time `db reset` is called.

**Comment by joeally** [comment]
I've realised that if I dump the `storage.buckets` and `storage.objects` tables into my `seed.sql` the files are retained across a database reset. However this would be of no help to someone who newly clones the repo or recreates the underlying docker containers. You'd still need to manually upload the files. I also don't know what happens if the rows exist in the postgres db but the corresponding files don't exist.

**Comment by plasmatech8** [comment]
It would be really nice if we had a `supabase/seed-storage/<bucket>/*` folder that we can use to automatically seed files.

(Would upload files after the `migrations/*` scripts, and preferably runs before `seed.sql` in case there is seed data that depends on it on the objects)

**Comment by robcurtis** [comment]
Is there any workaround for this? I'm currently facing this same problem.

**Comment by Donnerstagnacht** [comment]
I would be also interested in this feature or a workaround.

**Comment by mlawlerau** [comment]
I think this would be a useful feature

**Comment by logemann** [comment]
Dont know why this REAL problem is not tackled by supabase. I mean seeding storage files is equally important as seeding DB records. Guys...please upvote to get more attention to this....

**Comment by sweatybridge** [comment, marked as answer]
This is now supported in CLI beta by declaring your buckets in `supabase/config.toml`. For eg.

```toml
[storage.buckets.images]
public = false
file_size_limit = "50MiB"
allowed_mime_types = ["image/png", "image/jpeg"]
objects_path = "./path/to/images"
```

The declared buckets will be automatically seeded when you run the following commands
- `npx supabase@beta start`
- `npx supabase@beta db reset`

To seed a remote project, use `npx supabase@beta seed buckets --linked` in your CI/CD pipeline.

Note that it's not recommended to commit large binary files into your git repo (> 10MB). To seed large files, you want to download them separately to known `objects_path` before seeding.

**Reply by logemann** [reply]
great stuff

**Reply by xen-amar** [reply]
Some notes:
- Bucket to seed is specified in the configuration group name: `storage.buckets.<bucketName>`
- Object path (`objects_path`) seems to be relative to the `supabase/` subfolder
- Relevant CLI configuration docs section: https://supabase.com/docs/guides/local-development/cli/config#storage-config

**Reply by Lucas-Irvine** [reply]
@sweatybridge Is there a current solution for extending this to paths within a given bucket? ie it would be great to do

```
[storage.buckets.images.path.to.directory]
public = false
file_size_limit = "50MiB"
allowed_mime_types = ["image/png", "image/jpeg"]
objects_path = "./path/to/images"
```

For our use case, we upload files to dynamic paths within a given bucket using an object's id as the path so that certain files can be associated with different objects.

**Reply by thorwebdev** [reply]
TIL 🙌 This needs to be in the docs!

**Reply by sweatybridge** [reply]
@Lucas-Irvine we probably need to support a new config block for your use case. For eg.

```toml
[[storage.objects]]
bucket = "images"
key_prefix = "path/to/directory"
source = "./path/to/images"
cache_control = "3600s"
# content_type = "image/png"
```

It can be declared multiple times for different `path/to/directory` within the same bucket, and also allows customising object properties like `cache_control`.

**Reply by MorganFrary** [reply]
Is there anyway we can get this to support xml/svg files? Currently I'm receiving an invalid_mine_type error for text/xml but I'd expect "image/svg+xml" as a mime type to be supported.

**Reply by juliussneezer04** [reply]
> @sweatybridge Is there a current solution for extending this to paths within a given bucket? ie it would be great to do
> 
> ```
> [storage.buckets.images.path.to.directory]
> public = false
> file_size_limit = "50MiB"
> allowed_mime_types = ["image/png", "image/jpeg"]
> objects_path = "./path/to/images"
> ```
> 
> For our use case, we upload files to dynamic paths within a given bucket using an object's id as the path so that certain files can be associated with different objects.

@Lucas-Irvine  I managed to workaround this by making `objects_path` point to a directory containing my other subdirectories and it got seeded with the proper file structure hierarchy, i.e. doing
```
[storage.buckets.images]
public = false
file_size_limit = "50MiB"
allowed_mime_types = ["image/png", "image/jpeg"]
objects_path = "./images-bucket"
```

where `images-bucket` is a dir with structure like
```
|
| - path
        | - to
             | - directory
```
(pls excuse the shoddy diagram)

**Reply by disposableraft** [reply]
This is great! However, it looks like the database reset happens before `buckets.Run`, which means that `seeds.sql` cannot populate any file-specific data. 

```go
	// Reset postgres database because extensions (pg_cron, pg_net) require postgres
	if err := resetDatabase(ctx, version, fsys, options...); err != nil {
		return err
	}
	// Seed objects from supabase/buckets directory
	if resp, err := utils.Docker.ContainerInspect(ctx, utils.StorageId); err == nil {
		if resp.State.Health == nil || resp.State.Health.Status != types.Healthy {
			if err := start.WaitForHealthyService(ctx, 30*time.Second, utils.StorageId); err != nil {
				return err
			}
		}
		if err := buckets.Run(ctx, "", false, fsys); err != nil {
			return err
		}
	}
	branch := keys.GetGitBranch(fsys)
	fmt.Fprintln(os.Stderr, "Finished "+utils.Aqua("supabase db reset")+" on branch "+utils.Aqua(branch)+".")
	return nil
}
```

[link](https://github.com/supabase/cli/blob/af413835f7996c2ac0752438abfb0932a6280cf5/internal/db/reset/reset.go#L55-L73)

**Comment by keeandev** [comment]
One thing I would recommend is a .gitignore for files to not upload, or following the existing project gitignore, or file specific exclusions with a glob pattern. I was unable to seed because there were some hidden .DS_Store files in the directories I was trying to upload into a bucket, and my allowed mime types didn't accept them.

**Reply by keeandev** [reply]
Or even just don't upload files that aren't in the allowed mime types list, and provide a warning during seeding.
